File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        total = 0
        stack = []
        n = len(height)
        
        for i in range(n):
            while stack and height[stack[-1]] < height[i]:
                curr_height = height[stack.pop()]
                if not stack:
                    break
                bricks = min(height[i], height[stack[-1]]) - curr_height
                total += bricks * (i - stack[-1] - 1)
                
            stack.append(i)
        return total

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        // height = [0,1,0,2,1,0,1,3,2,1,2,1]
        // [3,2, 1]
        // 1 + 1 + 3 + 1
        stack<int>st;
        int n = height.size(), ans = 0;
        for (int i = 0; i < n; i++) {
            while (!st.empty() && height[st.top()] <= height[i]) {
                int cur = height[st.top()];
                st.pop();
                if (st.empty()) break;
                int high = min(height[i], height[st.top()]) - cur;
                ans += high * (i - st.top() - 1);
            }
            st.push(i);
        }
        return ans;
    }
};